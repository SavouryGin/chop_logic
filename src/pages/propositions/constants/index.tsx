import { GreekSymbol, LogicalSymbolHexCode, LogicalSymbolRawInput } from 'enums';

export * from './tab-lists';

export * from './symbols';

const constants = {
  logicalOperators: [
    LogicalSymbolRawInput.Implication,
    LogicalSymbolRawInput.Conjunction,
    LogicalSymbolRawInput.Disjunction,
    LogicalSymbolRawInput.Negation,
    LogicalSymbolRawInput.Equivalence,
  ],

  parentheses: [LogicalSymbolRawInput.OpenParenthesis, LogicalSymbolRawInput.CloseParenthesis],

  directProofsEditorTableColumns: [
    {
      field: 'step',
      title: { en: '#', ru: '№' },
    },
    {
      field: 'friendlyExpression',
      title: { en: 'Formula', ru: 'Формула' },
    },
    {
      field: 'comment',
      title: { en: 'Comment', ru: 'Коммент.' },
    },
  ],

  naturalProofsEditorTableColumns: [
    {
      field: 'step',
      title: { en: '#', ru: '№' },
    },
    {
      field: 'friendlyExpression',
      title: { en: 'Formula', ru: 'Формула' },
    },
    {
      field: 'comment',
      title: { en: 'Comment', ru: 'Коммент.' },
    },
  ],

  implicationCreationInitialValues: { firstVariable: '', secondVariable: '' },

  contradictionRealization: { firstVariable: '', secondVariable: '' },

  implicationDistributionInitialValues: { firstVariable: '', secondVariable: '', thirdVariable: '' },

  implicationCreationFormula: `${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ( ${GreekSymbol.Psi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Phi} )`,

  implicationDistributionFormula: `( ${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ( ${GreekSymbol.Psi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Chi} )) ${LogicalSymbolHexCode.Implication} (( ${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Psi} ) ${LogicalSymbolHexCode.Implication} ( ${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Chi} ))`,

  contradictionRealizationFormula: `( ${LogicalSymbolHexCode.Negation}${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Psi} ) ${LogicalSymbolHexCode.Implication} (( ${LogicalSymbolHexCode.Negation}${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${LogicalSymbolHexCode.Negation}${GreekSymbol.Psi} ) ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Phi} )`,
};

export default Object.freeze(constants);
