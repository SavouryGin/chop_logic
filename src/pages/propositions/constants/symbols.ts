import { LogicalSymbolHexCode, LogicalSymbolRawInput } from 'enums';
import { PropositionalSymbol } from 'types';

export const preparedSymbols: { [key in string]: PropositionalSymbol } = {
  openParenthesis: {
    input: LogicalSymbolRawInput.OpenParenthesis,
    representation: '(',
    type: 'parentheses',
    position: 0,
  },

  closeParenthesis: {
    input: LogicalSymbolRawInput.CloseParenthesis,
    representation: ')',
    type: 'parentheses',
    position: 0,
  },

  implication: {
    input: LogicalSymbolRawInput.Implication,
    representation: LogicalSymbolHexCode.Implication,
    type: 'operator',
    position: 0,
  },

  conjunction: {
    input: LogicalSymbolRawInput.Conjunction,
    representation: LogicalSymbolHexCode.Conjunction,
    type: 'operator',
    position: 0,
  },

  disjunction: {
    input: LogicalSymbolRawInput.Disjunction,
    representation: LogicalSymbolHexCode.Disjunction,
    type: 'operator',
    position: 0,
  },

  equivalence: {
    input: LogicalSymbolRawInput.Equivalence,
    representation: LogicalSymbolHexCode.Equivalence,
    type: 'operator',
    position: 0,
  },

  negation: {
    input: LogicalSymbolRawInput.Negation,
    representation: LogicalSymbolHexCode.Negation,
    type: 'operator',
    position: 0,
  },
};
