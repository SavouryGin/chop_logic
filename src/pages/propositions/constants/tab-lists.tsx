import AxiomSchemesForDirectPropositionProofs from 'pages/propositions/tabs/direct-proofs/axiom-schemes';
import DirectProofsEditor from 'pages/propositions/tabs/direct-proofs/editor';
import DirectProofsPropositionDefinitions from 'pages/propositions/tabs/direct-proofs/definitions';
import HeuristicsOfPropositions from 'pages/propositions/tabs/natural-proofs/heuristics';
import NaturalProofsEditor from 'pages/propositions/tabs/natural-proofs/editor';
import NaturalProofsPropositionDefinitions from 'pages/propositions/tabs/natural-proofs/definitions';
import React from 'react';
import RulesOfInference from 'pages/propositions/tabs/natural-proofs/rules';
import texts from 'texts/propositions/elements';
import { TabItem } from 'types';

export const propositionsDirectProofsTabs: TabItem[] = [
  {
    tabContent: <DirectProofsEditor />,
    tabTitle: texts.editor,
    tabId: 'direct-proofs-editor',
  },
  {
    tabContent: <DirectProofsPropositionDefinitions />,
    tabTitle: texts.definitions,
    tabId: 'direct-proofs-definitions',
  },
  {
    tabContent: <AxiomSchemesForDirectPropositionProofs />,
    tabTitle: texts.axioms,
    tabId: 'axiom-schemes-for-direct-proofs',
  },
];

export const propositionsNaturalProofsTabs: TabItem[] = [
  {
    tabContent: <NaturalProofsEditor />,
    tabTitle: texts.editor,
    tabId: 'natural-proofs-editor',
  },
  {
    tabContent: <NaturalProofsPropositionDefinitions />,
    tabTitle: texts.definitions,
    tabId: 'natural-proofs-definitions',
  },
  {
    tabContent: <RulesOfInference />,
    tabTitle: texts.rules,
    tabId: 'rules-of-inference',
  },
  {
    tabContent: <HeuristicsOfPropositions />,
    tabTitle: texts.heuristics,
    tabId: 'heuristics-of-propositions',
  },
];
