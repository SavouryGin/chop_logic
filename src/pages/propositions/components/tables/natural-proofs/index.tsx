import React, { useEffect, useState } from 'react';
import Table from 'components/table';
import constants from 'pages/propositions/constants';
import propositionsElementsTexts from 'texts/propositions/elements';
import { propositionsNPActions as actions } from 'store/propositions/natural-proofs';
import { npSelectors as selectors } from 'store/propositions/natural-proofs/selectors';
import { settingsSelectors } from 'store/settings/selectors';
import { useAppDispatch, useAppSelector } from 'hooks';

const NaturalProofsEditorTable = (): React.ReactElement => {
  const dispatch = useAppDispatch();
  const language = useAppSelector(settingsSelectors.language);
  const tableData = useAppSelector(selectors.tableData);
  const [selectedIds, setSelectedIds] = useState<string[]>(useAppSelector(selectors.selectedIds));
  const takeSelectedIds = (ids: string[]) => setSelectedIds(ids);

  useEffect(() => {
    dispatch(actions.setSelectedIds(selectedIds));
  }, [selectedIds]);

  const noStepsFiller = <div className='natural-proofs-editor__filler'>{propositionsElementsTexts.fillerNaturalText[language]}</div>;

  return (
    <div className='natural-proofs-editor__table'>
      <Table
        columns={constants.naturalProofsEditorTableColumns}
        data={tableData}
        hasCheckboxColumn
        passSelectedIds={takeSelectedIds}
        id='propositions-np-editor-table'
      />
      {tableData.length === 0 && noStepsFiller}
    </div>
  );
};

export default NaturalProofsEditorTable;
