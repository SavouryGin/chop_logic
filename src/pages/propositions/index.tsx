import React from 'react';
import TabList from 'components/tab-list';
import formatClassName from 'helpers/formatters/format-class-name';
import DirectProofsEditor from './elements/direct-proofs-editor';
import NaturalProofsEditor from './elements/natural-proofs-editor';
import DefinitionsOfPropositions from './elements/definitions-of-propositions';
import HeuristicsOfPropositions from './elements/heuristics-of-propositions';
import RulesOfInference from './elements/rules-of-inference';
import { TabItem } from 'types';
import { Icon } from 'enums';
import { pageTitle, tabTitles } from 'assets/texts/propositions';
import { useAppSelector } from 'hooks';
import { settingsSelectors } from 'store/settings/selectors';

import './styles.scss';

export const propositionsTabs: TabItem[] = [
  {
    tabContent: <DirectProofsEditor />,
    tabTitle: tabTitles.direct,
    tabId: 'direct-proofs-editor',
  },
  {
    tabContent: <NaturalProofsEditor />,
    tabTitle: tabTitles.natural,
    tabId: 'natural-proofs-editor',
  },
  {
    tabContent: <DefinitionsOfPropositions />,
    tabTitle: tabTitles.definitions,
    tabId: 'definitions-of-propositions',
  },
  {
    tabContent: <HeuristicsOfPropositions />,
    tabTitle: tabTitles.heuristics,
    tabId: 'heuristics-of-propositions',
  },
  {
    tabContent: <RulesOfInference />,
    tabTitle: tabTitles.rules,
    tabId: 'rules-of-inference',
  },
];

function Propositions(): React.ReactElement {
  const language = useAppSelector(settingsSelectors.getLanguage);
  return (
    <article className='propositions'>
      <h2 className={formatClassName(['propositions__title', Icon.Propositions])}>{pageTitle[language]}</h2>
      <TabList tabs={propositionsTabs} className='propositions__tabs' />
    </article>
  );
}

export default Propositions;
