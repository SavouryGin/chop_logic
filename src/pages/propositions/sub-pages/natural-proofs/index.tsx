import React from 'react';
import TabList from 'components/tab-list';
import formatClass from 'helpers/formatters/format-class-name';
import { Icon } from 'enums';
import { propositionsNaturalProofsTabs } from 'pages/propositions/constants';
import { settingsSelectors } from 'store/settings/selectors';
import { titles } from 'texts/propositions';
import { useAppSelector } from 'hooks';
import './styles.scss';

const PropositionsNaturalProofs = () => {
  const language = useAppSelector(settingsSelectors.getLanguage);

  return (
    <article className='propositions-natural-proofs'>
      <h2
        className={formatClass(['propositions-natural-proofs__title', Icon.Propositions])}
      >{`${titles.page[language]} > ${titles.natural[language]}`}</h2>
      <TabList tabs={propositionsNaturalProofsTabs} className='propositions-natural-proofs__tabs' />
    </article>
  );
};

export default PropositionsNaturalProofs;
