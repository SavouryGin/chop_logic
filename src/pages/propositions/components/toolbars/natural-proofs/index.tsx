import Button from 'components/controls/button';
import React from 'react';
import { ButtonID } from 'enums';
import { npActions as actions } from 'store/propositions/natural-proofs';
import { npSelectors as selectors } from 'store/propositions/natural-proofs/selectors';
import { soundPlayer } from 'helpers/sounds';
import {
  useAppDispatch,
  useAppSelector,
  useIsAndEliminationPossible,
  useIsEquivalenceEliminationPossible,
  useIsEquivalenceIntroductionPossible,
  useIsImpliesEliminationForNPPossible,
  useIsImpliesIntroductionPossible,
  useIsNotEliminationPossible,
  useIsNotIntroductionPossible,
  useIsOrEliminationPossible,
  useIsPremisePossible,
} from 'hooks';

const NaturalProofsEditorToolbar = (): React.ReactElement => {
  const dispatch = useAppDispatch();
  const isPremiseDisabled = !useIsPremisePossible();
  const selectedIds = useAppSelector(selectors.selectedIds);
  const tableDataLength = useAppSelector(selectors.tableDataLength);
  const isReplacerDisabled = tableDataLength === 0;
  const isReiterationDisabled = selectedIds.length !== 1;
  const isDeleteDisabled = selectedIds.length === 0;
  const isOrIntroductionDisabled = selectedIds.length === 0;
  const isOrEliminationDisabled = !useIsOrEliminationPossible(selectedIds);
  const isAndIntroductionDisabled = !selectedIds.length;
  const isAndEliminationDisabled = !useIsAndEliminationPossible(selectedIds);
  const isNotIntroductionDisabled = !useIsNotIntroductionPossible(selectedIds);
  const isNotEliminationDisabled = !useIsNotEliminationPossible(selectedIds);
  const isEquivIntroductionDisabled = !useIsEquivalenceIntroductionPossible(selectedIds);
  const isEquivEliminationDisabled = !useIsEquivalenceEliminationPossible(selectedIds);
  const isImpliesEliminationDisabled = !useIsImpliesEliminationForNPPossible(selectedIds);
  const isImpliesIntroductionDisabled = !useIsImpliesIntroductionPossible(selectedIds);

  const openPremise = () => {
    dispatch(actions.setUpFlag({ flag: 'isPremiseOpened', value: true }));
  };

  const openAssumption = () => {
    dispatch(actions.setUpFlag({ flag: 'isAssumptionOpened', value: true }));
  };

  const deleteSteps = () => {
    dispatch(actions.deleteSteps({ isConfirmed: false }));
  };

  const reiterateStep = () => {
    dispatch(actions.reiterateStep());
  };

  const openOrIntroduction = () => {
    dispatch(actions.setUpFlag({ flag: 'isOrIntroductionFormOpened', value: true }));
  };

  const eliminateDisjunction = () => {
    dispatch(actions.eliminateDisjunction());
  };

  const createConjunction = () => {
    dispatch(actions.createConjunction());
  };

  const eliminateConjunction = () => {
    dispatch(actions.eliminateConjunction());
  };

  const createNegation = () => {
    dispatch(actions.createNegation());
  };

  const eliminateNegation = () => {
    dispatch(actions.eliminateNegation());
  };

  const createEquivalence = () => {
    dispatch(actions.createEquivalence());
  };

  const eliminateEquivalence = () => {
    dispatch(actions.eliminateEquivalence());
  };

  const eliminateImplication = () => {
    dispatch(actions.eliminateImplication());
  };

  const createImplication = () => {
    dispatch(actions.createImplication());
  };

  const openShortcut = () => {
    dispatch(actions.setUpFlag({ flag: 'isShortcutOpened', value: true }));
  };

  const openReplacer = () => {
    dispatch(actions.setUpFlag({ flag: 'isReplacerFormOpened', value: true }));
  };

  return (
    <div className='natural-proofs-editor__toolbar'>
      <Button buttonId={ButtonID.Premise} sound={soundPlayer.keyboard} view='large' onClick={openPremise} isDisabled={isPremiseDisabled} />
      <Button
        buttonId={ButtonID.Reiteration}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={reiterateStep}
        isDisabled={isReiterationDisabled}
      />
      <Button
        buttonId={ButtonID.Replace}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={openReplacer}
        isDisabled={isReplacerDisabled}
      />
      <Button
        buttonId={ButtonID.NotIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={createNegation}
        isDisabled={isNotIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.AndIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={createConjunction}
        isDisabled={isAndIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.OrIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={openOrIntroduction}
        isDisabled={isOrIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.ImpliesIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={createImplication}
        isDisabled={isImpliesIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.EquivIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        isDisabled={isEquivIntroductionDisabled}
        onClick={createEquivalence}
      />
      <Button buttonId={ButtonID.Assumption} sound={soundPlayer.keyboard} view='large' onClick={openAssumption} />
      <Button buttonId={ButtonID.Delete} sound={soundPlayer.keyboard} view='large' onClick={deleteSteps} isDisabled={isDeleteDisabled} />
      <Button buttonId={ButtonID.Shortcut} sound={soundPlayer.keyboard} view='large' onClick={openShortcut} />
      <Button
        buttonId={ButtonID.NotElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateNegation}
        isDisabled={isNotEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.AndElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateConjunction}
        isDisabled={isAndEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.OrElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateDisjunction}
        isDisabled={isOrEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.ImpliesElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateImplication}
        isDisabled={isImpliesEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.EquivElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateEquivalence}
        isDisabled={isEquivEliminationDisabled}
      />
    </div>
  );
};

export default NaturalProofsEditorToolbar;
