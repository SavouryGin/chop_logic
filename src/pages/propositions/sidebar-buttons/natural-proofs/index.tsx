import Button from 'components/controls/button';
import ConfirmDeleteProofStepsPopup from 'pages/propositions/components/forms/confirm-delete';
import ModalWindow from 'components/modal-window';
import React from 'react';
import { ButtonID, Icon } from 'enums';
import { propositionsNPActions as actions } from 'store/propositions/natural-proofs';
import { propositionsNPSelectors as selectors } from 'store/propositions/natural-proofs/selectors';
import { settingsSelectors } from 'store/settings/selectors';
import { uiElementTexts } from 'texts';
import { useAppDispatch, useAppSelector } from 'hooks';

const PropositionsNPSidebarButtons = ({ isVisible }: { isVisible: boolean }): React.ReactElement | null => {
  const dispatch = useAppDispatch();
  const language = useAppSelector(settingsSelectors.language);
  const tableDataLength = useAppSelector(selectors.tableDataLength);
  const selectedIds = useAppSelector(selectors.selectedIds);
  const clipboardData = useAppSelector(selectors.clipboardData);
  const dependencies = useAppSelector(selectors.dependentItems);
  const isConfirmCutPopupOpened = useAppSelector(selectors.isConfirmCutPopupOpened);

  if (!isVisible) {
    return null;
  }

  const isExportToXMLDisabled = !tableDataLength;
  const isCopyDisabled = !selectedIds.length;
  const isPasteDisabled = !clipboardData.length;
  const isCutDisabled = !selectedIds.length;

  const onClickExportToXML = () => {
    dispatch(actions.exportToXML());
  };

  const openFileInputForm = () => {
    dispatch(actions.setUpFlag({ flag: 'isUserFileFormVisible', value: true }));
  };

  const onCopySteps = () => {
    dispatch(actions.copySteps());
  };

  const onPasteSteps = () => {
    dispatch(actions.pasteSteps());
  };

  const onCutSteps = () => {
    dispatch(actions.cutSteps({ isConfirmed: false }));
  };

  const closeCutSteps = () => {
    dispatch(actions.setUpFlag({ flag: 'isConfirmCutPopupOpened', value: false }));
    dispatch(actions.setDependentItems([]));
  };

  const confirmCutSteps = () => {
    dispatch(actions.setUpFlag({ flag: 'isConfirmCutPopupOpened', value: false }));
    dispatch(actions.cutSteps({ isConfirmed: true }));
  };

  return (
    <>
      <li>
        <Button
          buttonId={ButtonID.ExportXML}
          icon={Icon.ExportXML}
          isDisabled={isExportToXMLDisabled}
          onClick={onClickExportToXML}
          view='large'
        />
      </li>
      <li>
        <Button buttonId={ButtonID.ImportXML} icon={Icon.ImportXML} view='large' onClick={openFileInputForm} />
      </li>
      <li>
        <Button buttonId={ButtonID.CopyProof} icon={Icon.Copy} view='large' onClick={onCopySteps} isDisabled={isCopyDisabled} />
      </li>
      <li>
        <Button buttonId={ButtonID.CutProof} icon={Icon.Cut} view='large' onClick={onCutSteps} isDisabled={isCutDisabled} />
      </li>
      <li>
        <Button buttonId={ButtonID.PasteProof} icon={Icon.Paste} view='large' onClick={onPasteSteps} isDisabled={isPasteDisabled} />
      </li>
      <ModalWindow
        isOpened={isConfirmCutPopupOpened}
        onClose={closeCutSteps}
        title={uiElementTexts.confirmation[language]}
        content={<ConfirmDeleteProofStepsPopup onConfirm={confirmCutSteps} dependencies={dependencies} />}
      />
    </>
  );
};

export default PropositionsNPSidebarButtons;
