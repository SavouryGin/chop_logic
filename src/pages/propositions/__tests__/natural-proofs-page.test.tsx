import PropositionsNaturalProofs from 'pages/propositions/sub-pages/natural-proofs';
import React from 'react';
import renderWithRedux from 'helpers/test-utils/render-with-redux';
import { combineReducers } from '@reduxjs/toolkit';
import { propositionsNP, propositionsNPInitialState } from 'store/propositions/natural-proofs/slice';
import { propositionsNaturalProofsTabs } from 'pages/propositions/constants';
import { screen } from '@testing-library/react';
import { settingsInitialState, settingsSlice } from 'store/settings/slice';
import { titles } from 'texts/propositions';

describe('Natural Proofs page tests:', () => {
  const mockedReducer = combineReducers({
    settings: settingsSlice.reducer,
    propositionsNP: propositionsNP.reducer,
  });

  const mockedState = {
    settings: settingsInitialState,
    propositionsNP: propositionsNPInitialState,
  };

  beforeEach(() => {
    renderWithRedux(<PropositionsNaturalProofs />, mockedReducer, mockedState);
  });

  it('renders the article element', () => {
    expect(screen.getByRole('article')).toBeInTheDocument();
    expect(screen.getByRole('article')).toHaveClass('propositions-natural-proofs');
  });

  it('displays the heading', () => {
    expect(screen.getByRole('heading')).toBeInTheDocument();
    expect(screen.getByRole('heading')).toHaveTextContent(titles.page.en);
  });

  it('renders the correct number of tabs', () => {
    expect(screen.getAllByRole('tab')).toHaveLength(propositionsNaturalProofsTabs.length);
  });

  it('the tabs have correct titles', () => {
    for (const item of propositionsNaturalProofsTabs) {
      expect(screen.getByTestId(item.tabId)).toHaveTextContent(item.tabTitle.en);
    }
  });

  it('should match the snapshot', () => {
    const { asFragment } = renderWithRedux(<PropositionsNaturalProofs />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });
});
