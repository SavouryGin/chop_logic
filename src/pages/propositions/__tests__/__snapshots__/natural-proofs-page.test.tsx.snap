// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Natural Proofs page tests: should match the snapshot 1`] = `
<DocumentFragment>
  <article
    class="propositions-natural-proofs"
  >
    <h2
      class="propositions-natural-proofs__title chop-icon__cube"
    >
      Propositions &gt; Natural Proofs
    </h2>
    <div
      class="tab-list propositions-natural-proofs__tabs"
    >
      <div
        class="tab-list__tabs"
        role="tablist"
      >
        <h3
          aria-selected="true"
          class="tab-list__tab tab-list__tab_active"
          data-testid="natural-proofs-editor"
          id="natural-proofs-editor"
          role="tab"
        >
          Editor
        </h3>
        <h3
          aria-selected="false"
          class="tab-list__tab"
          data-testid="natural-proofs-definitions"
          id="natural-proofs-definitions"
          role="tab"
        >
          Definitions
        </h3>
        <h3
          aria-selected="false"
          class="tab-list__tab"
          data-testid="rules-of-inference"
          id="rules-of-inference"
          role="tab"
        >
          Rules
        </h3>
        <h3
          aria-selected="false"
          class="tab-list__tab"
          data-testid="heuristics-of-propositions"
          id="heuristics-of-propositions"
          role="tab"
        >
          Heuristics
        </h3>
      </div>
      <section
        aria-labelledby="natural-proofs-editor"
        class="tab-list__content"
        role="tabpanel"
      >
        <div
          class="natural-proofs-editor"
        >
          <div
            class="natural-proofs-editor__table"
          >
            <table
              class="table"
            >
              <thead
                class="table__head"
              >
                <tr
                  class="table__row"
                >
                  <th
                    class="table__cell"
                  >
                    <div
                      class="checkbox-input table__checkbox checkbox-input_disabled"
                    >
                      <input
                        class="checkbox-input__default"
                        disabled=""
                        id="select_all_in_propositions-np-editor-table"
                        name="select_all_in_propositions-np-editor-table"
                        type="checkbox"
                      />
                      <label
                        class="label checkbox-input__label chop-icon__checkbox-unchecked"
                        data-testid="label-test-id"
                        for="select_all_in_propositions-np-editor-table"
                      >
                        <span
                          class="label__text"
                        />
                      </label>
                    </div>
                  </th>
                  <th
                    class="table__cell table__cell_heading"
                  >
                    #
                  </th>
                  <th
                    class="table__cell table__cell_heading"
                  >
                    Formula
                  </th>
                  <th
                    class="table__cell table__cell_heading"
                  >
                    Comment
                  </th>
                </tr>
              </thead>
              <tbody
                class="table__body"
              />
            </table>
            <div
              class="natural-proofs-editor__filler"
            >
              Enter a premise or an assumption to start the proof
            </div>
          </div>
          <div
            class="natural-proofs-editor__toolbar"
          >
            <button
              class="button button_large"
              id="button_id_8"
              title="Enter premise"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_8"
              >
                Premise
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_9"
              title="Reiterate proof step"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_9"
              >
                Reiteration
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_10"
              title="Replace symbol"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_10"
              >
                Replace
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_17"
              title="Negation Introduction"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_17"
              >
                ¬ I
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_19"
              title="Conjunction Introduction"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_19"
              >
                ∧ I
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_21"
              title="Disjunction Introduction"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_21"
              >
                ∨ I
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_23"
              title="Implication Introduction"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_23"
              >
                ⇒ I
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_25"
              title="Equivalence Introduction"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_25"
              >
                ⇔ I
              </span>
            </button>
            <button
              class="button button_large"
              id="button_id_16"
              title="Assumption"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_16"
              >
                Assumption
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_11"
              title="Delete proof step"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_11"
              >
                Delete
              </span>
            </button>
            <button
              class="button button_large"
              id="button_id_27"
              title="Shortcut"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_27"
              >
                Shortcut
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_18"
              title="Negation Elimination"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_18"
              >
                ¬ E
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_20"
              title="Conjunction Elimination"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_20"
              >
                ∧ E
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_22"
              title="Disjunction Elimination"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_22"
              >
                ∨ E
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_24"
              title="Implication Elimination"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_24"
              >
                ⇒ E
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_26"
              title="Equivalence Elimination"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_26"
              >
                ⇔ E
              </span>
            </button>
          </div>
        </div>
      </section>
    </div>
  </article>
</DocumentFragment>
`;
